cmd/cloudFoundryDeploy.go

```diff
@@ -0,0 +1,855 @@
+	log.Entry().Infof("cfManifestVariables: '%v'", config.ManifestVariables)
+	log.Entry().Infof("cfManifestVariablesFiles: '%v'", config.ManifestVariablesFiles)
+	log.Entry().Infof("cfdeployDockerImage: '%s'", config.DeployDockerImage)
+	log.Entry().Infof("smokeTestScript: '%s'", config.SmokeTestScript)
+
+	additionalEnvironment := []string{
+		"STATUS_CODE=" + strconv.FormatInt(int64(config.SmokeTestStatusCode), 10),
+	}
+
+	if len(config.DockerPassword) > 0 {
+		additionalEnvironment = append(additionalEnvironment, "CF_DOCKER_PASSWORD="+config.DockerPassword)
+	}
+
+	myDeployConfig := deployConfig{
+		DeployCommand:   deployCommand,
+		DeployOptions:   deployOptions,
+		AppName:         config.AppName,
+		ManifestFile:    config.Manifest,
+		SmokeTestScript: smokeTestScript,
+	}
+
+	log.Entry().Infof("DeployConfig: %v", myDeployConfig)
+
+	//return nil
+	return deployCfNative(myDeployConfig, config, additionalEnvironment, command)
+}
+
+func deployCfNative(deployConfig deployConfig, config *cloudFoundryDeployOptions, additionalEnvironment []string, cmd command.ExecRunner) error {
+
+	deployStatement := []string{
+		deployConfig.DeployCommand,
+	}
+
+	if len(deployConfig.AppName) > 0 {
+		deployStatement = append(deployStatement, deployConfig.AppName)
+	}
+
+	if len(deployConfig.DeployOptions) > 0 {
+		deployStatement = append(deployStatement, deployConfig.DeployOptions...)
+	}
+
+	if len(deployConfig.ManifestFile) > 0 {
+		deployStatement = append(deployStatement, "-f")
+		deployStatement = append(deployStatement, deployConfig.ManifestFile)
+	}
+
+	if len(config.DeployDockerImage) > 0 && config.DeployType != "blue-green" {
+		deployStatement = append(deployStatement, "--docker-image", config.DeployDockerImage)
+	}
+
+	if len(config.DockerUsername) > 0 && config.DeployType != "blue-green" {
+		deployStatement = append(deployStatement, "--docker-username", config.DockerUsername)
+	}
+
+	if len(deployConfig.SmokeTestScript) > 0 {
+		deployStatement = append(deployStatement, deployConfig.SmokeTestScript...)
+	}
+
+	if len(config.CfNativeDeployParameters) > 0 {
+		deployStatement = append(deployStatement, strings.FieldsFunc(config.CfNativeDeployParameters, func(c rune) bool {
+			return c == ' '
+		})...)
+	}
+
+	stopOldAppIfRunning := func(_cmd command.ExecRunner) error {
+
+		if config.KeepOldInstance && config.DeployType == "blue-green" {
+			oldAppName := deployConfig.AppName + "-old"
+			_ = oldAppName
+
+			var buff bytes.Buffer
+
+			oldOut := _cmd.Stdout
+			_ = oldOut
+			_cmd.Stdout(&buff)
+
+			defer func() {
+				_cmd.Stdout(log.Writer())
+			}()
+
+			err := _cmd.RunExecutable("cf", "stop", oldAppName)
+
+			if err != nil {
+
+				cfStopLog := buff.String()
+
+				if !strings.Contains(cfStopLog, oldAppName+" not found") {
+					return fmt.Errorf("Could not stop application %s. Error: %s", oldAppName, cfStopLog)
+				}
+				log.Entry().Infof("Cannot stop application '%s': %s", oldAppName, cfStopLog)
```

Comment:
No. This log message appears if the app can be found, but cannot be stopped. The "not found" case results in an `error`. In this case we do not reach that line. In groovy this was simply silently ignored.

I'm asking me: Wouldn't it be better to raise an error here instead of writing a log message?

Reply:
Are you sure? In line 334 you return an error, if the log does **not** contain the sequence "<oldAppName> not found". So when we reach line 336, the log contains these lines. So in both cases: The app could not been found, or it was found, but could not be stopped, the log will say "Cannot stop" or "Could not stop". While in line 336, what it should say is that the app was not found.

Reply:
> I'm asking me: Wouldn't it be better to raise an error here instead of writing a log message?

I think it is better to not fail the pipeline if the old app cannot be found. I am probably not aware of all the details, but if you usually configure the step to stop the old instance, you could only successfully run the pipeline if there already is an instance started by other means.

Reply:
got it, it is a normal use case that the app is not already present. Log message adjusted.